sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1277,6 +1346,34 @@ static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
 		 */
 		if (codec_info->dais[dai_index].rtd_init_done)
 			continue;
+
+		/*
+		 * Add card controls and dapm widgets for the first codec dai.
+		 * The controls and widgets will be used for all codec dais.
+		 */
+		if (i == 0) {
+			const struct snd_soc_dapm_widget *w = codec_info->dais[dai_index].widgets;
+			const struct snd_kcontrol_new *c = codec_info->dais[dai_index].controls;
```

Comment:
Minor nitpick but I am not a huge fan of single letter variable names outside of loop counters, but willing to look the other way if everyone else is happy.

Also a thought (more of discussion point than definitely saying we should do this), but would there be mileage in using the generic widgets/controls based on dai_type if the link doesn't define custom ones? That way you every new dai_link wouldn't need to add links to the generic ones. Really it would also be nice if all new code used the generic ones unless there was a very clear reason not to. Only slight difficulties I see are the realtek mic codecs that currently define no widgets although that seems weird since they presumably do have a mic associated and the mock up codecs. Although in all those cases the dangling widgets will be redundant but not cause any issues.

Reply:
> Minor nitpick but I am not a huge fan of single letter variable names outside of loop counters, but willing to look the other way if everyone else is happy.

I use a single letter is for the 100-character limitation. I am happy if someone can suggest a better way to fit the limitation.

> 
> Also a thought (more of discussion point than definitely saying we should do this), but would there be mileage in using the generic widgets/controls based on dai_type if the link doesn't define custom ones? That way you every new dai_link wouldn't need to add links to the generic ones. Really it would also be nice if all new code used the generic ones unless there was a very clear reason not to. Only slight difficulties I see are the realtek mic codecs that currently define no widgets although that seems weird since they presumably do have a mic associated and the mock up codecs. Although in all those cases the dangling widgets will be redundant but not cause any issues.

Yes, I thought about it, too. However, I saw not all codecs eg. cs35l56 add their own controls. So, I don't add controls/widgets by default for now.

Reply:
updated.